core/src/apps/ontology/serialize.py

```diff
@@ -0,0 +1,133 @@
+from trezor.messages.OntologyOntIdAddAttributes import OntologyOntIdAddAttributes
+from trezor.messages.OntologyOntIdRegister import OntologyOntIdRegister
+from trezor.messages.OntologyTransaction import OntologyTransaction
+from trezor.messages.OntologyTransfer import OntologyTransfer
+from trezor.messages.OntologyTxAttribute import OntologyTxAttribute
+from trezor.messages.OntologyWithdrawOng import OntologyWithdrawOng
+
+from . import const as Const, writer
+from .helpers import get_bytes_from_address
+from .sc.native_builder import ParamStruct, build_native_call
+
+
+def serialize_tx(tx: OntologyTransaction, payload: bytes, hw):
+    writer.write_byte(hw, tx.version)
+    writer.write_byte(hw, tx.type)
+    writer.write_uint32(hw, tx.nonce)
+    writer.write_uint64(hw, tx.gas_price)
+    writer.write_uint64(hw, tx.gas_limit)
+    payer = get_bytes_from_address(tx.payer)
+    writer.write_bytes(hw, payer)
+
+    writer.write_bytes(hw, payload)
+
+    attributes = tx.tx_attributes
+    writer.write_varint(hw, len(attributes))
+
+    if attributes is not None:
+        for attribute in attributes:
+            _serialize_tx_attribute(hw, attribute)
+
+
+def serialize_transfer(transfer: OntologyTransfer):
+    """
+    Serializes transaction with specified transfer as payload
+    """
+    return _serialize_transfer_payload(transfer)
+
+
+def serialize_withdraw_ong(withdraw_ong: OntologyWithdrawOng):
+    """
+    Serializes transaction with specified withdraw Ong as payload
+    """
+    return _serialize_withdraw_ong_payload(withdraw_ong)
+
+
+def serialize_ont_id_register(register: OntologyOntIdRegister):
+    """
+    Serializes transaction with specified ONT ID registration as payload
+    """
+    return _serialize_ont_id_register_payload(register)
+
+
+def serialize_ont_id_add_attributes(add: OntologyOntIdAddAttributes):
+    """
+    Serializes transaction with specified ONT ID attributes adding as payload
+    """
+    return _serialize_ont_id_add_attributes_payload(add)
+
+
+def _serialize_tx_attribute(ret: bytearray, attribute: OntologyTxAttribute):
+    writer.write_byte(ret, attribute.usage)
+
+    if attribute.data is not None:
+        writer.write_bytes_with_length(ret, attribute.data)
+
+
+def _serialize_transfer_payload(transfer: OntologyTransfer):
+    from_address = get_bytes_from_address(transfer.from_address)
+    to_address = get_bytes_from_address(transfer.to_address)
+    amount = transfer.amount
+    contract = ""
+
+    if transfer.asset == OntologyAsset.ONT:
+        contract = Const.ONT_CONTRACT
+    else:
+        contract = Const.ONG_CONTRACT
+
+    struct = ParamStruct([from_address, to_address, amount])
+    native_call = build_native_call("transfer", [[struct]], contract)
+
+    # 9 is the maximum possible length of the uvarint prefix
+    # ret = bytearray(len(native_call) + 9)
+    ret = bytearray()
+    writer.write_bytes_with_length(ret, native_call)
+    return bytes(ret)
```

Comment:
I tried to do this but tests stop passing. I suspect it's because the writer used in serialize_tx is trezor.utils.HashWriter. I can still return the native_call and move L84-85 to serialize_tx, does that make sense?

Reply:
after discussion we determined that we need to keep this here as-is

(it is because the returned payload isn't only used in `serialize_tx`, but also sent back to the host as the `payload` attribute.

OTOH maybe the host doesn't actually need that?)

Reply:
After further discussion we decided to drop the payload attribute from protobuf and refactor this as discussed earlier.